rslib/src/scheduler/filtered/custom_study.rs

```diff
@@ -0,0 +1,226 @@
+            }
+            CustomStudyValue::ReviewAheadDays(days) => {
+                self.create_custom_study_deck(ahead_config(current_deck.human_name(), days))
+            }
+            CustomStudyValue::PreviewDays(days) => {
+                self.create_custom_study_deck(preview_config(current_deck.human_name(), days))
+            }
+            CustomStudyValue::Cram(cram) => {
+                self.create_custom_study_deck(cram_config(current_deck.human_name(), cram)?)
+            }
+        }
+    }
+
+    /// Reuse existing one or create new one if missing.
+    /// Guaranteed to be a filtered deck.
+    fn create_custom_study_deck(&mut self, config: FilteredDeck) -> Result<()> {
+        let mut id = DeckId(0);
+        let human_name = self.tr.custom_study_custom_study_session().to_string();
+
+        if let Some(did) = self.get_deck_id(&human_name)? {
+            if !self
+                .get_deck(did)?
+                .ok_or(AnkiError::NotFound)?
+                .is_filtered()
+            {
+                return Err(CustomStudyError::ExistingDeck.into());
+            }
+            id = did;
+        }
+
+        self.add_or_update_filtered_deck_inner(FilteredDeckForUpdate {
+            id,
+            human_name,
+            config,
+        })
+        .map(|_| ())
+        .map_err(|err| {
+            if err == AnkiError::FilteredDeckError(FilteredDeckError::SearchReturnedNoCards) {
+                CustomStudyError::NoMatchingCards.into()
+            } else {
+                err
+            }
+        })
+    }
+}
+
+fn custom_study_config(
+    deck_name: String,
+    reschedule: bool,
+    nodes: Vec<Node>,
+    order: FilteredSearchOrder,
+    limit: Option<u32>,
+) -> FilteredDeck {
+    let search = concatenate_searches(
+        BoolSeparator::And,
+        nodes,
+        Node::Search(SearchNode::Deck(deck_name)),
+    );
+    FilteredDeck {
+        reschedule,
+        search_terms: vec![FilteredSearchTerm {
+            search,
+            limit: limit.unwrap_or(99_999),
+            order: order as i32,
+        }],
+        delays: vec![],
+        preview_delay: 10,
+    }
+}
+
+fn forgot_config(deck_name: String, days: u32) -> FilteredDeck {
+    let nodes = vec![Node::Search(SearchNode::Rated {
+        days,
+        ease: RatingKind::AnswerButton(1),
+    })];
+    custom_study_config(deck_name, false, nodes, FilteredSearchOrder::Random, None)
+}
+
+fn ahead_config(deck_name: String, days: u32) -> FilteredDeck {
+    let nodes = vec![Node::Search(SearchNode::Property {
+        operator: "<=".to_string(),
+        kind: PropertyKind::Due(days as i32),
+    })];
+    custom_study_config(deck_name, true, nodes, FilteredSearchOrder::Due, None)
+}
+
+fn preview_config(deck_name: String, days: u32) -> FilteredDeck {
+    let nodes = vec![
+        Node::Search(SearchNode::State(StateKind::New)),
+        Node::And,
+        Node::Search(SearchNode::AddedInDays(days)),
+    ];
```

Comment:
The existing macros and TryIntoSearch trait could make this a bit easier:

```diff
@@ -130,11 +132,11 @@ fn ahead_config(deck_name: String, days: u32) -> FilteredDeck {
 }

 fn preview_config(deck_name: String, days: u32) -> FilteredDeck {
-    let nodes = vec![
-        Node::Search(SearchNode::State(StateKind::New)),
-        Node::And,
-        Node::Search(SearchNode::AddedInDays(days)),
-    ];
+    let nodes = match_all![
+        SearchNode::State(StateKind::New),
+        SearchNode::AddedInDays(days)
+    ]
+    .into_node_list();
     custom_study_config(
         deck_name,
         false,
```

If custom_study_config() were reworked to avoid the concatenate and additional deck argument, we could just add the deck to the list, and make things a bit simpler again:

```diff
diff --git a/rslib/src/scheduler/filtered/custom_study.rs b/rslib/src/scheduler/filtered/custom_study.rs
index 269b3ce16..e428d3678 100644
--- a/rslib/src/scheduler/filtered/custom_study.rs
+++ b/rslib/src/scheduler/filtered/custom_study.rs
@@ -13,7 +13,8 @@ use crate::{
     error::{CustomStudyError, FilteredDeckError},
     prelude::*,
     search::{
-        concatenate_searches, BoolSeparator, Node, PropertyKind, RatingKind, SearchNode, StateKind,
+        concatenate_searches, writer::write_node, BoolSeparator, Node, PropertyKind, RatingKind,
+        SearchNode, StateKind,
     },
     text::escape_anki_wildcards_for_search_node,
 };
@@ -113,6 +114,24 @@ fn custom_study_config(
     }
 }
 
+fn custom_study_config_example(
+    reschedule: bool,
+    node: Node,
+    order: FilteredSearchOrder,
+    limit: Option<u32>,
+) -> FilteredDeck {
+    FilteredDeck {
+        reschedule,
+        search_terms: vec![FilteredSearchTerm {
+            search: write_node(&node),
+            limit: limit.unwrap_or(99_999),
+            order: order as i32,
+        }],
+        delays: vec![],
+        preview_delay: 10,
+    }
+}
+
 fn forgot_config(deck_name: String, days: u32) -> FilteredDeck {
     let nodes = vec![Node::Search(SearchNode::Rated {
         days,
@@ -130,18 +149,12 @@ fn ahead_config(deck_name: String, days: u32) -> FilteredDeck {
 }
 
 fn preview_config(deck_name: String, days: u32) -> FilteredDeck {
-    let nodes = vec![
-        Node::Search(SearchNode::State(StateKind::New)),
-        Node::And,
-        Node::Search(SearchNode::AddedInDays(days)),
+    let node = match_all![
+        SearchNode::State(StateKind::New),
+        SearchNode::AddedInDays(days),
+        SearchNode::Deck(deck_name),
     ];
-    custom_study_config(
-        deck_name,
-        false,
-        nodes,
-        FilteredSearchOrder::OldestReviewedFirst,
-        None,
-    )
+    custom_study_config_example(false, node, FilteredSearchOrder::OldestReviewedFirst, None)
 }
 
 macro_rules! state_node {
diff --git a/rslib/src/search/writer.rs b/rslib/src/search/writer.rs
index 0b700d915..a87f333ac 100644
--- a/rslib/src/search/writer.rs
+++ b/rslib/src/search/writer.rs
@@ -69,7 +69,7 @@ pub fn write_nodes(nodes: &[Node]) -> String {
     nodes.iter().map(write_node).collect()
 }
 
-fn write_node(node: &Node) -> String {
+pub fn write_node(node: &Node) -> String {
     use Node::*;
     match node {
         And => " ".to_string(),
```

Maybe instead of making write_node() public, we should implement a write() method on Node.

Reply:
Thanks, I was missing these tools. I think I'll expand on them in a separate file.